src/mito2/src/request.rs

```diff
@@ -248,46 +256,67 @@ impl WriteRequest {
     pub(crate) fn fill_missing_columns(&mut self, metadata: &RegionMetadata) -> Result<()> {
         debug_assert_eq!(self.region_id, metadata.region_id);
 
+        let mut columns_to_fill = vec![];
         for column in &metadata.column_metadatas {
             if !self.name_to_index.contains_key(&column.column_schema.name) {
-                self.fill_column(column)?;
+                columns_to_fill.push(column);
             }
         }
+        self.fill_columns(columns_to_fill)?;
 
         Ok(())
     }
 
-    /// Fills default value for specific `column`.
-    fn fill_column(&mut self, column: &ColumnMetadata) -> Result<()> {
-        // Need to add a default value for this column.
-        let proto_value = self.column_default_value(column)?;
+    /// Checks the schema and fill missing columns.
+    pub(crate) fn maybe_fill_missing_columns(&mut self, metadata: &RegionMetadata) -> Result<()> {
+        if let Err(e) = self.check_schema(metadata) {
+            if e.is_fill_default() {
+                // TODO(yingwen): Add metrics for this case.
+                // We need to fill default value. The write request may be a request
+                // sent before changing the schema.
+                self.fill_missing_columns(metadata)?;
+            } else {
+                return Err(e);
+            }
+        }
 
-        if proto_value.value_data.is_none() {
-            return Ok(());
+        Ok(())
+    }
+
+    /// Fills default value for specific `columns`.
+    fn fill_columns(&mut self, columns: Vec<&ColumnMetadata>) -> Result<()> {
+        let mut default_values = Vec::with_capacity(columns.len());
+        let mut columns_to_fill = Vec::with_capacity(columns.len());
+        for column in columns {
+            let default_value = self.column_default_value(column)?;
+            if default_value.value_data.is_some() {
+                default_values.push(default_value);
+                columns_to_fill.push(column);
+            }
         }
 
-        // Insert default value to each row.
         for row in &mut self.rows.rows {
-            row.values.push(proto_value.clone());
+            row.values.extend(default_values.clone());
```

Comment:
non-blocking: Maybe we can only clone the items.
```suggestion
            row.values.extend(default_values.iter().cloned());
```